### PR TITLE
CgltfImporter

### PR DESCRIPTION
**Pull request by pezcode**
👋 Here's the initial version of `CgltfImporter` to import glTF 2.0 files using [cgltf](https://github.com/jkuhlmann/cgltf).

As far as features are concerned, it's a drop-in replacement for `TinyGltfImporter`, but there are a few minor changes in behaviour (see below). Large parts of its code and tests were re-used and adapted for cgltf, making some parts more trivial than expected.

There aren't any major TODOs, only a few things I'd appreciate some feedback on, all marked with `@todo` in the code (some are kept from `TinyGltfImporter` code). So unless CI breaks I won't be changing anything until the first review round.

### Noticable improvements

*Compared to `TinyGltfImporter`, using VS 2019, Release build*:

- Compile time: 5s -> 3s
- Binary size: 430kb -> 220kb

### Subtle differences that may impact users switching from `TinyGltfImporter`

-  Buffers are loaded on demand and references/data are cached. This means the callback policy is now `LoadPermanent` and returned views on resources have to stay valid until the importer is closed.
- `WEIGHTS` and `JOINTS` are imported as a single custom attribute with multiple sets, mimicking behaviour of `AssimpImporter` and preparing for https://github.com/mosra/magnum/pull/441
- [Percent-encoded URIs](https://www.khronos.org/registry/glTF/specs/2.0/glTF-2.0.html#uris) are now decoded, `TinyGltfImporter` ignored percent encoding.

### New features

- Support for [KHR_texture_basisu](https://github.com/KhronosGroup/glTF/blob/main/extensions/2.0/Khronos/KHR_texture_basisu/README.md) and [MSFT_texture_dds](https://github.com/KhronosGroup/glTF/blob/main/extensions/2.0/Vendor/MSFT_texture_dds/README.md) (in addition to the already existing `GOOGLE_texture_basis`)
- `asset.version` is checked against the supported glTF version (2.0)
- a few extra warnings about invalid mesh attribute names

### CODE REVIEW

**Comment by mosra** [comment]
Oooo, wonderful! :tada: Looking forward to diving into the code.

Wait, the tinygltf plugin compiles in just 5 seconds for you? Here I always have to wait about half a minute for a compile + link, longer than all other plugins combined... The binary size reduction is awesome.

> `KHR_texture_basisu`

Wow, also? :star_struck: I suppose there's some bits missing in `KtxImporter` to delegate to `BasisImporter`, it doesn't work just yet, right?

**Comment by pezcode** [comment]
> I suppose there's some bits missing in `KtxImporter` to delegate to `BasisImporter`, it doesn't work just yet, right?

Yeah, that needs some extra wiring in `KtxImporter`.

As for CI failures: the Emscripten one I'm not sure about, seems like a recent issue with sourceforge? The AddressSanitizer error is interesting, I'll have to see if that's a genuine bug but it has to be fixed upstream either way. It doesn't like overlapping `strcpy` ranges in coded added in https://github.com/jkuhlmann/cgltf/pull/165

File: src/MagnumPlugins/CgltfImporter/CgltfImporter.cpp [hunk 1]
```diff
@@ -0,0 +1,2562 @@
...
+            continue;
+
+        const cgltf_attribute& attribute = primitive.attributes[a];
+
+        const Containers::StringView nameString{attribute.name};
+        /* See the comment in doOpenData() for why we do this */
+        const Containers::StringView semantic = attribute.type != cgltf_attribute_type_invalid ?
+            nameString.partition('_')[0] : nameString;
+
+        if(attribute.type != cgltf_attribute_type_invalid) {
```
Comment (mosra) [inline src/MagnumPlugins/CgltfImporter/CgltfImporter.cpp:1524]: Took me a while to realize what these warnings are for, wasn't clear from the test either :) Comment, perhaps?

```suggestion
        /* Numbered attributes are expected to be contiguous (COLORS_0,
           COLORS_1...). If not, print a warning, because in the MeshData they
           will appear as contiguous. */
        if(attribute.type != cgltf_attribute_type_invalid) {
```

File: src/MagnumPlugins/CgltfImporter/CgltfImporter.cpp [hunk 2]
```diff
@@ -0,0 +1,2562 @@
...
+            if(attribute.type != lastAttribute.type) {
+                if(attribute.index != 0)
+                    Warning{} << "Trade::CgltfImporter::mesh(): expected attribute" << nameString << "to start at index 0";
+            } else if(attribute.index != lastAttribute.index + 1)
+                Warning{} << "Trade::CgltfImporter::mesh(): expected attribute" << nameString << "to continue at index" << lastAttribute.index + 1;
+        }
```
Comment (mosra) [inline src/MagnumPlugins/CgltfImporter/CgltfImporter.cpp:1530]: What about 

```suggestion
        if(attribute.type != cgltf_attribute_type_invalid && ((attribute.type != lastAttribute.type && attribute.index != 0) || attribute.index != lastAttribute.index + 1))
            Warning{} << "Trade::CgltfImporter::mesh(): found attribute" << nameString << "but there was no" << Utility::formatString("{}_{}", semantic, attribute.index - 1);
```

which would print e.g. `Trade::CgltfImporter::mesh(): found attribute COLORS_3 but there was no COLORS_2`? I can understand that warning a lot better.

File: src/MagnumPlugins/CgltfImporter/CgltfImporter.cpp [hunk 3]
```diff
@@ -0,0 +1,2562 @@
...
+        } else if(nameString == configuration().value("objectIdAttribute")) {
+            name = MeshAttribute::ObjectId;
+
+            if(accessor->type != cgltf_type_scalar) {
+                Error{} << "Trade::CgltfImporter::mesh(): unexpected object ID type" << accessor->type;
+                return Containers::NullOpt;
+            }
+
+            /* The glTF spec says that "Application-specific attribute semantics
+               MUST NOT use unsigned int component type" but I'm not sure what
```
Comment (mosra) [inline src/MagnumPlugins/CgltfImporter/CgltfImporter.cpp:1693]: What the hell :eyes:

File: src/MagnumPlugins/CgltfImporter/Test/CgltfImporterTest.cpp [hunk 4]
```diff
@@ -0,0 +1,4435 @@
...
+}
+
+void CgltfImporterTest::meshInvalidIndexComponentType() {
+    Containers::Pointer<AbstractImporter> importer = _manager.instantiate("CgltfImporter");
+
+    /* cgltf_validate checks index accessor types, but only if the mesh has any
+       attributes. The test case in mesh-invalid.gltf has no attributes, so it
+       only tests our own check. This one tests the result of cgltf_validate.
+       This is all a bit shitty since the check is now duplicated and occurs in
+       different functions (openData vs. mesh) and with different error text. */
```
Comment (mosra) [inline src/MagnumPlugins/CgltfImporter/Test/CgltfImporterTest.cpp:3126]: Ah well. I was just wondering why there's the additional `mesh-invalid-*.gltf` files. Together with my other comment about underwhelming error messages, would it make sense to reimplement `cgltf_validate()` on our side to have the handling more consistent? Basically we'd have to always look at what it does and then add the missing checks on our side (especially for extensions cgltf doesn't support yet), but instead if we'd be handling everything then it's much easier to ensure all cases are handled properly.

**Review by mosra — COMMENTED** [review]
I went through hopefully everything, what's left is looking at the coverage and checking if there are some untested areas. 

- The Emscripten CI failure is worked around in a7a8b43696ac91bc37e2a41528e743a56acc073f, so if you rebase then it should go through.
- The ASan failure about overlapping `strcpy()` is valid I guess (would happen with e.g. `\\ab` already, copying `ab` over `\a`), and same as with `memcpy()` it's not allowed. Not sure if there's any standard `strmove()` equivalent to `memmove()` tho, I suppose you'd need to get the string size first and then use the `mem*()` APIs instead?

File: src/MagnumPlugins/CgltfImporter/CgltfImporter.cpp [hunk 5]
```diff
@@ -0,0 +1,2562 @@
...
+    const auto found = decodedStrings.find(str.data());
+    if(found != decodedStrings.end())
+        return found->second;
+
+    /* The input string can be UTF-8 encoded but we can use a byte search here
+       since all multi-byte UTF-8 characters have the high bit set and '\\'
+       doesn't, so this will only match single-byte ASCII characters. */
+    Containers::StringView escape = str.find('\\');
+    /* No escaped sequence found. If the view is null-terminated (all strings
+       in cgltf_data should be), this stores a non-owning String. */
```
[the remark this answers is not in the document]
Comment (pezcode) [reply src/MagnumPlugins/CgltfImporter/CgltfImporter.cpp:427]: It does allocate a null-terminated copy, yes. And there's no convenient way to get the original pointer to json memory that I can see. There are two alternatives to the current implementation:

1. Decode all strings we could possibly need later. Not a fan of imposing that overhead at import time.
2. Same as now, but decode strings in-place on cgltf-allocated strings (or even `delete[]` and replace). Saves some memory but we'd still have to track what's already decoded, maybe with an `unordered_set`.

I went with the `unordered_map` because it felt cleaner to abstract (not really true), but I'm starting to like option 2 a bit more, if just for the memory savings. Opinions?

File: src/MagnumPlugins/CgltfImporter/CgltfImporter.cpp [hunk 6]
```diff
@@ -0,0 +1,2562 @@
...
+    const std::size_t start = escape.data() - str.data();
+
+    const std::size_t size = str.size();
+    Containers::Array<char> decoded{size + 1};
+    Utility::copy(str, decoded.prefix(size));
+    decoded.back() = '\0';
+
+    cgltf_decode_string(decoded.data() + start);
+
+    const std::size_t decodedSize = strlen(decoded.data() + start) + start;
```
[the remark this answers is not in the document]
Comment (pezcode) [reply src/MagnumPlugins/CgltfImporter/CgltfImporter.cpp:441]: I'll try to submit that upstream, no harm in returning that info 👍

File: src/MagnumPlugins/CgltfImporter/CgltfImporter.cpp [hunk 7]
```diff
@@ -0,0 +1,2562 @@
...
+            case cgltf_result_io_error:
+                CORRADE_INTERNAL_ASSERT_UNREACHABLE(); /* LCOV_EXCL_LINE */
+            /* This should never happen, it means we passed a nullptr somewhere */
+            case cgltf_result_invalid_options:
+                CORRADE_INTERNAL_ASSERT_UNREACHABLE(); /* LCOV_EXCL_LINE */
+            case cgltf_result_success:
+                CORRADE_INTERNAL_ASSERT_UNREACHABLE(); /* LCOV_EXCL_LINE */
+        }
+
+        return "unknown error";
```
[the remark this answers is not in the document]
Comment (pezcode) [reply src/MagnumPlugins/CgltfImporter/CgltfImporter.cpp:551]: They're all in the switch, so that can indeed be an assert instead

File: src/MagnumPlugins/CgltfImporter/CgltfImporter.cpp [hunk 8]
```diff
@@ -0,0 +1,2562 @@
...
+
+Containers::Optional<TextureData> CgltfImporter::doTexture(const UnsignedInt id) {
+    const cgltf_texture& tex = _d->data->textures[id];
+
+    UnsignedInt imageId = ~0u;
+
+    /* Various extensions, they override the standard image */
+    if(tex.has_basisu && tex.basisu_image) {
+        /* KHR_texture_basisu. Allows the usage of mimeType image/ktx2 but only
+           explicitly talks about KTX2 with Basis compression. We don't care
```
[the remark this answers is not in the document]
Comment (pezcode) [reply src/MagnumPlugins/CgltfImporter/CgltfImporter.cpp:2301]: For simplicity I would've preferred a generic "everything" KTX extension, but I can see why they split up the texture formats this way.

File: src/MagnumPlugins/CgltfImporter/CgltfImporter.cpp [hunk 9]
```diff
@@ -0,0 +1,2562 @@
...
+    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+    THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+    FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+    DEALINGS IN THE SOFTWARE.
+*/
+
+#include "CgltfImporter.h"
+
+#include <algorithm>
```
[the remark this answers is not in the document]
Comment (pezcode) [reply src/MagnumPlugins/CgltfImporter/CgltfImporter.cpp:29]: It's needed for `std::stable_sort` for sorting mesh attributes. I'd assume the small array optimization in there is faster than using `unordered_map` for less than ~5 attributes in an average mesh?

File: src/MagnumPlugins/CgltfImporter/Test/CgltfImporterTest.cpp [hunk 10]
```diff
@@ -0,0 +1,4435 @@
...
+void CgltfImporterTest::animationOutOfBounds() {
+    auto&& data = AnimationOutOfBoundsData[testCaseInstanceId()];
+    setTestCaseDescription(data.name);
+
+    Containers::Pointer<AbstractImporter> importer = _manager.instantiate("CgltfImporter");
+
+    std::ostringstream out;
+    Error redirectError{&out};
+    CORRADE_VERIFY(!importer->openFile(Utility::Directory::join(TINYGLTFIMPORTER_TEST_DIR, data.file)));
+    CORRADE_COMPARE(out.str(), "Trade::CgltfImporter::openData(): error opening file: invalid glTF\n");
```
[the remark this answers is not in the document]
Comment (pezcode) [reply src/MagnumPlugins/CgltfImporter/Test/CgltfImporterTest.cpp:1168]: Would it make sense to expand on the generic error message a bit? Just to give people a hint on what *might be wrong*... Something like:

> "error opening file: invalid glTF, usually caused by invalid indices or missing required attributes"

File: src/MagnumPlugins/CgltfImporter/Test/CgltfImporterTest.cpp (hunk 4 above)
Comment (pezcode) [reply src/MagnumPlugins/CgltfImporter/Test/CgltfImporterTest.cpp:3126]: Doing our own validation is probably the better option, indeed. Saves us from silly edge cases like above, adds useful (and consistent) error messages and moves some checks to only run when really needed. I initially dismissed it because of code duplication, but I think the advantages overweigh :eyes:

File: src/MagnumPlugins/CgltfImporter/CgltfImporter.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (mosra) [reply src/MagnumPlugins/CgltfImporter/CgltfImporter.cpp:427]: I was just wondering why the design decision in cgltf / jsmn went like that, why it doesn't go all the way and unescape as well, or why on the other hand it doesn't just keep pointing to the original data if it's unescaped anyway. Doing a copy just to make them null-terminated feels silly tbh :)

Your implementation on the plugin side is okay, I wouldn't change that, for now at least. If it ever becomes a bottleneck, I'd instead look into fixing this directly in cgltf or jsmn (not sure which of the two is responsible).

File: src/MagnumPlugins/CgltfImporter/CgltfImporter.cpp [hunk 11]
```diff
@@ -0,0 +1,2562 @@
...
+    /* Copy file content. We need to keep the data around for .glb binary blobs
+       and extension data which cgltf stores as pointers into the original
+       memory passed to cgltf_parse. */
+    _d->fileData = Containers::Array<char>{data.size()};
+    Utility::copy(data, _d->fileData);
+
+    /* Auto-detect glb/gltf */
+    _d->options.type = cgltf_file_type::cgltf_file_type_invalid;
+    /* Determine json token count to allocate (by parsing twice) */
+    _d->options.json_token_count = 0;
```
[the remark this answers is not in the document]
Comment (mosra) [reply src/MagnumPlugins/CgltfImporter/CgltfImporter.cpp:498]: Unix systems should allow that with some `mmap()` variant I think, Windows has `virtual_alloc()` ([relevant blog post on Our Machinery](https://ourmachinery.com/post/virtual-memory-tricks/)). Possibly some sandboxing / UWP APIs could impose limitations. WebAssembly doesn't have a concept of virtual memory yet, but [it's being worked on](https://github.com/WebAssembly/design/issues/1397). The future is bright! :)

File: src/MagnumPlugins/CgltfImporter/CgltfImporter.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (mosra) [reply src/MagnumPlugins/CgltfImporter/CgltfImporter.cpp:29]: Ah, good point, can you then add that as a comment so it's clear why the include is needed?

```suggestion
#include <algorithm> /* std::stable_sort() */
```

File: src/MagnumPlugins/CgltfImporter/Test/CgltfImporterTest.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (mosra) [reply src/MagnumPlugins/CgltfImporter/Test/CgltfImporterTest.cpp:1168]: That would make sense indeed :+1: 

I'm afraid this could become a friction point, I remember people were already complaining that tinygltf's json.hpp only prints byte position but not line number when reporting JSON syntax errors, which makes debugging kinda annoying. If it becomes a problem, it'd probably have to be fixed on cgltf's side, not much to do here.

[the remark this answers is not in the document]
Comment (pezcode) [reply src/MagnumPlugins/CgltfImporter/Test/CgltfImporterTest.cpp:1168]: We could probably link to the [glTF validator](https://github.khronos.org/glTF-Validator) in the docs? Just as a least-effort hint for now

File: src/MagnumPlugins/CgltfImporter/CgltfImporter.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (pezcode) [reply src/MagnumPlugins/CgltfImporter/CgltfImporter.cpp:427]: jsmn doesn't perform any allocation, so this is cgltf's doing. Makes it a bit easier to fix things upstream if we ever need that, but I suspect removing null-terminated strings wouldn't be easy to get merged in any case.

**Comment by pezcode** [comment]
I addressed everything except moving validation to the plugin side, I'll work on that tomorrow.

I'll send PRs for the two cgltf changes related to string decoding, but those might take a few days to get merged. Should I push the changes locally for now so tests pass and you can see test coverage?

File: src/MagnumPlugins/CgltfImporter/CgltfImporter.cpp [hunk 12]
```diff
@@ -0,0 +1,2595 @@
...
+
+/* Decode percent-encoded characters in URIs:
+   https://datatracker.ietf.org/doc/html/rfc3986#section-2.1 */
+std::string decodeUri(Containers::StringView uri) {
+    std::string decoded = uri;
+    cgltf_decode_uri(&decoded[0]);
+    decoded.resize(std::strlen(decoded.data()));
+    /** @todo Replace backslash with forward slash? AssimpImporter normalizes
+        paths that way */
+    return decoded;
```
[the remark this answers is not in the document]
Comment (mosra) [reply src/MagnumPlugins/CgltfImporter/CgltfImporter.cpp:176]: How likely is it for glTF files with backslashes to exist? FBX is historically a Windows-centric format, so backlashes there don't surprise me at all, but for modern glTF exporters I would expect a modicum of platform awareness at least :)

So I'd say until we actually encouter malformed glTFs with backlashes, I don't see a reason for dealing with those. In addition, backslashes are a valid character in a filename on most platforms except Windows, so if we'd want to do it _correctly_ it'd be much more complex than a find&replace, and then I wonder if it shouldn't be dealt with at some lower level.

File: src/MagnumPlugins/CgltfImporter/Test/CgltfImporterTest.cpp [hunk 13]
```diff
@@ -0,0 +1,4416 @@
...
+    {"unexpected position type", "mesh-invalid.gltf", "unexpected POSITION type 2"},
+    {"unsupported position component type", "mesh-invalid.gltf", "unsupported POSITION component type unnormalized VertexFormat::UnsignedInt"},
+    {"unexpected normal type", "mesh-invalid.gltf", "unexpected NORMAL type 2"},
+    {"unsupported normal component type", "mesh-invalid.gltf", "unsupported NORMAL component type unnormalized VertexFormat::UnsignedInt"},
+    {"unexpected tangent type", "mesh-invalid.gltf", "unexpected TANGENT type 3"},
+    {"unsupported tangent component type", "mesh-invalid.gltf", "unsupported TANGENT component type unnormalized VertexFormat::Byte"},
+    {"unexpected texcoord type", "mesh-invalid.gltf", "unexpected TEXCOORD type 3"},
+    {"unsupported texcoord component type", "mesh-invalid.gltf", "unsupported TEXCOORD component type normalized VertexFormat::UnsignedInt"},
+    {"unexpected color type", "mesh-invalid.gltf", "unexpected COLOR type 2"},
+    {"unsupported color component type", "mesh-invalid.gltf", "unsupported COLOR component type unnormalized VertexFormat::Byte"},
```
[the remark this answers is not in the document]
Comment (mosra) [reply src/MagnumPlugins/CgltfImporter/Test/CgltfImporterTest.cpp:366]: Good point. Do something like `unsupported COLOR component type unnormalized VertexFormat::Byte (58062)`? A number alone is useless unless one cross-checks with the glTF spec (the GL headers have those in hex, so those don't help either), but it's still useful so one can for example grep for it in the file and change appropriately.

File: src/MagnumPlugins/CgltfImporter/Test/CgltfImporterTest.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (mosra) [reply src/MagnumPlugins/CgltfImporter/Test/CgltfImporterTest.cpp]: Good idea -- mention the possibility of vague errors in the behavior/limitations docs and add a link to the validator.

**Comment by pezcode** [comment]
Ah, this is missing test coverage for invalid input to `gltf*TypeName()`. I'll add a test for that via `doAnimation`, that should also cover the two missing types in the switch. Going to push that with whatever changes are left after your review.

**Comment by mosra** [comment]
Yep :)

**Comment by pezcode** [comment]
I squashed everything down, but I kept [a branch](https://github.com/pezcode/magnum-plugins/tree/cgltf-importer-commit-history) in my fork with the full commit history just in case 👀

**Comment by mosra** [comment]
Ahh, playing with the new plugin feels great, everything exceeds expectations. In my case it's the following (native builds use the `scenedata` magnum branch, which may add some extra overhead due to all extra compat interfaces):

- Compile time GCC Debug: **19 -> 6 seconds**
- Compile time GCC Release: **14 -> 6.5 seconds**
- Plugin binary size GCC Debug: **10 MB -> 2.5 MB**
- Plugin binary size GCC Release: **696 kB -> 277 kB**
- Plugin binary size Emscripten Release: **4.2 MB -> 1.2 MB**
- Size of the Emscripten `magnum-viewer` example (`ports` branch, without the bundled scene): **914 kB -> 727 kB** 
  - For comparison the barebones triangle example that uses almost nothing from Magnum is 313 kB

I blame the leftover `std::string`s and `std::vector`s in the plugin manager interfaces that get always inlined, without those the binary sizes and compile times could go even less.

Looked at import time using `magnum-sceneconverter --info --profile`, CgltfImporter never gets slower than TinyGltfImporter but I suppose most of the time was spent in reading the file and copying the data a few extra times, and I only tested on a smallish 7.5 MB `Buggy.glb` from the glTF samples repo. Once I take care of the extra copies with mosra/magnum#240 (which CgltfImporter finally unblocks, yay!), it has the potential of outperforming everything else.

For memory consumption (`valgrind --tool=callgrind`) of the same tool, I see CgltfImporter peak memory usage being 16 MB down from 17.2 MB with TinyGltfImporter. Given that there's no other change in Magnum internals that would avoid having the file present in memory twice (which I really have to fix now), I find this quite great -- JSON parser overhead is less than half of the original. With a change that would reuse the memory allocated in `openFile()` instead of making a copy in `doOpenData()`, the peak memory consumption would be another 7.5 MB less, so **17.2 -> 8.5 MB** :tada:
